test/box-luatest/gh_8802_box_session_push_deprecation_test.lua

```diff
@@ -0,0 +1,94 @@
+local server = require('luatest.server')
+local t = require('luatest')
+
+local WARNING_PATTERN = 'W> box%.session%.push is deprecated. ' ..
+                        'Consider using box%.broadcast instead%.$'
+
+local g = t.group('box_session_push_deprecation', t.helpers.matrix({
+    api = {'lua', 'c'},
+    compat = {'old', 'new'},
+}))
+
+g.before_all(function(cg)
+    cg.server = server:new()
+end)
+
+g.before_each(function(cg)
+    cg.server:start()
+    cg.server:exec(function(params)
+        local func
+        if params.api == 'lua' then
+            func = box.session.push
+        elseif params.api == 'c' then
+            local ffi = require('ffi')
+            local msgpack = require('msgpack')
+            ffi.cdef([[
+                int box_session_push(const char *data, const char *data_end);
+            ]])
+            func = function(value)
+                local data = msgpack.encode(value)
+                if ffi.C.box_session_push(
+                        ffi.cast('const char *', data),
+                        ffi.cast('const char *', data) + #data) ~= 0 then
+                    return nil, box.error.last()
+                end
+                return true
+            end
+        end
+        t.assert(func)
+        rawset(_G, 'box_session_push', func)
+    end, {cg.params})
+end)
+
+g.after_each(function(cg)
+    cg.server:stop()
+end)
+
+g.after_all(function(cg)
+    cg.server:drop()
+end)
+
+g.test = function(cg)
+    cg.server:exec(function(params)
+        local compat = require('compat')
+        compat.box_session_push_deprecation = params.compat
+    end, {cg.params})
+
+    t.assert_not(cg.server:grep_log(WARNING_PATTERN),
+                 'No deprecation warning in the log initially')
+
+    -- Test box.session.push. It shouldn't work with the new behavior.
+    -- The warning should be logged in any case.
+    cg.server:exec(function()
+        local compat = require('compat')
+        local box_session_push = rawget(_G, 'box_session_push')
+        local ok, err = box_session_push('foo')
```

Comment:
Shouldn't we raise `ER_UNSUPPORTED` instead of returning it? Otherwise it may be easy to miss.

I know, all the other errors are returned as `nil, err`, but this one is specific and maybe it should hit a caller even if (s)he doesn't bother with checking for errors from `box.session.push()`.

What do you think?

Reply:
Actually, I agree with you. This complicates the code but it should be better for the users.

<details><summary>diff</summary>

```diff
diff --git a/src/box/box.cc b/src/box/box.cc
index d14558dcc7bd..2a8c57028f05 100644
--- a/src/box/box.cc
+++ b/src/box/box.cc
@@ -4000,6 +4000,8 @@ box_session_push(const char *data, const char *data_end)
 	struct session *session = current_session();
 	if (session == NULL)
 		return -1;
+	if (session_push_check_deprecation() != 0)
+		return -1;
 	struct port_msgpack port;
 	struct port *base = (struct port *)&port;
 	port_msgpack_create(base, data, data_end - data);
diff --git a/src/box/lua/session.c b/src/box/lua/session.c
index a90027262b0c..2777169c45fe 100644
--- a/src/box/lua/session.c
+++ b/src/box/lua/session.c
@@ -374,6 +374,8 @@ lbox_session_push(struct lua_State *L)
 	struct session *session = current_session();
 	if (lua_gettop(L) != 1)
 		return luaL_error(L, "Usage: box.session.push(data)");
+	if (session_push_check_deprecation() != 0)
+		return luaT_error(L);
 	struct port port;
 	port_lua_create(&port, L);
 	if (session_push(session, &port) != 0)
diff --git a/src/box/session.c b/src/box/session.c
index 4195bb2967b5..b9bf8e3ee00e 100644
--- a/src/box/session.c
+++ b/src/box/session.c
@@ -501,7 +501,7 @@ static bool box_session_push_is_disabled = true;
 TWEAK_BOOL(box_session_push_is_disabled);
 
 int
-session_push(struct session *session, struct port *port)
+session_push_check_deprecation(void)
 {
 	say_warn_once("box.session.push is deprecated. "
 		      "Consider using box.broadcast instead.");
@@ -509,7 +509,7 @@ session_push(struct session *session, struct port *port)
 		diag_set(ClientError, ER_DEPRECATED, "box.session.push");
 		return -1;
 	}
-	return session->vtab->push(session, port);
+	return 0;
 }
 
 int
diff --git a/src/box/session.h b/src/box/session.h
index a2ccd6f9dc5c..6e02e62afc32 100644
--- a/src/box/session.h
+++ b/src/box/session.h
@@ -403,9 +403,19 @@ access_check_universe_object(user_access_t access,
 			     enum schema_object_type object_type,
 			     const char *object_name);
 
-/** See session_vtab::push. */
+/**
+ * This function is called by public API wrappers around session push.
+ * It logs a deprecation warning. If session push is disabled, it also
+ * sets diag and returns -1.
+ */
 int
-session_push(struct session *session, struct port *port);
+session_push_check_deprecation(void);
+
+static inline int
+session_push(struct session *session, struct port *port)
+{
+	return session->vtab->push(session, port);
+}
 
 static inline int
 session_fd(struct session *session)
diff --git a/test/box-luatest/gh_8802_box_session_push_deprecation_test.lua b/test/box-luatest/gh_8802_box_session_push_deprecation_test.lua
index f689e8173823..804499ade56e 100644
--- a/test/box-luatest/gh_8802_box_session_push_deprecation_test.lua
+++ b/test/box-luatest/gh_8802_box_session_push_deprecation_test.lua
@@ -30,9 +30,8 @@ g.before_each(function(cg)
                 if ffi.C.box_session_push(
                         ffi.cast('const char *', data),
                         ffi.cast('const char *', data) + #data) ~= 0 then
-                    return nil, box.error.last()
+                    box.error()
                 end
-                return true
             end
         end
         t.assert(func)
@@ -62,13 +61,13 @@ g.test = function(cg)
     cg.server:exec(function()
         local compat = require('compat')
         local box_session_push = rawget(_G, 'box_session_push')
-        local ok, err = box_session_push('foo')
-        t.assert_equals(ok, compat.box_session_push_deprecation:is_old() or nil,
+        local ok, err = pcall(box_session_push, 'foo')
+        t.assert_equals(ok, compat.box_session_push_deprecation:is_old(),
                         'box.session.push status')
-        t.assert_equals(err and tostring(err),
-                        compat.box_session_push_deprecation:is_new() and
-                            'box.session.push is deprecated' or nil,
-                        'box.session.push error')
+        if not ok then
+            t.assert_equals(tostring(err), 'box.session.push is deprecated',
+                            'box.session.push error')
+        end
     end)
     t.assert(cg.server:grep_log(WARNING_PATTERN),
              'Deprecation warning is logged')
@@ -81,13 +80,13 @@ g.test = function(cg)
         end
         local compat = require('compat')
         local box_session_push = rawget(_G, 'box_session_push')
-        local ok, err = box_session_push('foo')
-        t.assert_equals(ok, compat.box_session_push_deprecation:is_old() or nil,
+        local ok, err = pcall(box_session_push, 'foo')
+        t.assert_equals(ok, compat.box_session_push_deprecation:is_old(),
                         'box.session.push status')
-        t.assert_equals(err and tostring(err),
-                        compat.box_session_push_deprecation:is_new() and
-                            'box.session.push is deprecated' or nil,
-                        'box.session.push error')
+        if not ok then
+            t.assert_equals(tostring(err), 'box.session.push is deprecated',
+                            'box.session.push error')
+        end
     end)
     t.assert_not(cg.server:grep_log(WARNING_PATTERN, 5000),
                  'Deprecation warning is logged only once')
```

</details>